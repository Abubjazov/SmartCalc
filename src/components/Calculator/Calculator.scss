@import '../../variables';

.calc-widget {
	margin-top: 70px;
	padding-top: 20px;

	h1 {
		font-size: 21px;
	}

	a {
		border-radius: 20px;
		padding: 12px;
		height: 40px;
		text-align: center;
		font-size: 18px;
		line-height: 18px;
		background-color: $main-color;
		color: $white;
		text-decoration: none;
		transition: all 0.1s linear;
	}

	input {
		height: 25px;
		width: 260px;
		border: none;
		border-bottom: 1px solid $deep-gray;
		outline: none;
		background: none;
		font-size: 20px;
		font-weight: 300;

		&:focus {
			border-bottom: 2px solid $main-color;
		}
	}

	p {
		position: absolute;
		top: 11vh;
		right: 2vw;
		font-size: 18px;
		font-weight: 300;
	}
}
