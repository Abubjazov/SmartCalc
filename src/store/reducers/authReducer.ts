import { AuthAction, AuthActionTypes, AuthState } from '../../interfaces'

const authDataString = localStorage.getItem('smartcalc_token')
const authData = authDataString ? JSON.parse(authDataString) : null

export const initialState: AuthState = {
	token: authData?.token || null,
	name: authData?.name || '',
	email: authData?.email || '',
	status: 'waiting',
	error: null,
}

export const authReducer = (
	state: AuthState = initialState,
	action: AuthAction
): AuthState => {
	switch (action.type) {
		case AuthActionTypes.FETCH_TOKEN:
			return {
				...state,
				status: 'loading',
				error: null,
			}

		case AuthActionTypes.FETCH_TOKEN_SUCCESS:
			return {
				...state,
				status: 'waiting',
				token: action.payload.token,
				name: action.payload.name,
				email: action.payload.email,
			}

		case AuthActionTypes.FETCH_TOKEN_ERROR:
			return {
				...state,
				status: 'error',
				error: action.payload,
			}

		case AuthActionTypes.DELETE_TOKEN:
			return {
				...state,
				status: 'loading',
				error: null,
			}

		case AuthActionTypes.DELETE_TOKEN_SUCCESS:
			return {
				...state,
				status: 'waiting',
				token: null,
				name: '',
				email: '',
			}

		case AuthActionTypes.DELETE_TOKEN_ERROR:
			return {
				...state,
				status: 'error',
				error: action.payload,
			}

		default:
			return state
	}
}
