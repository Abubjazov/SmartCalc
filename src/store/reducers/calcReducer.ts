import { nanoid } from 'nanoid'

import { CalcAction, CalcActionTypes, CalcState } from '../../interfaces'
import { matchCheck, returnInputItems, returnSummary } from '../../utils/utils'

export const initialState: CalcState = {
	items: [],
	inputItems: [
		{ key: nanoid(), value: '' },
		{ key: nanoid(), value: '' },
	],
	sortedItems: [],
	sortDirection: false,
	searchString: '',
	status: 'waiting',
	step: 1,
	summary: 0,
	error: null,
}

export const calcReducer = (
	state: CalcState = initialState,
	action: CalcAction
): CalcState => {
	switch (action.type) {
		case CalcActionTypes.FETCH_CURRENT_STATE:
			return { ...state, status: 'loading', error: null }

		case CalcActionTypes.FETCH_CURRENT_STATE_SUCCESS:
			return {
				...state,
				status: 'waiting',
				items: action.payload.items,
				sortedItems: action.payload.items,
				step: action.payload.step,
				inputItems: returnInputItems(action.payload.items),
				summary: returnSummary(action.payload.items),
			}

		case CalcActionTypes.FETCH_CURRENT_STATE_ERROR:
			return {
				...state,
				status: 'error',
				error: action.payload,
			}

		case CalcActionTypes.SWITCH_TO_CONFIRM:
			return { ...state, status: 'loading', error: null }

		case CalcActionTypes.SWITCH_TO_CONFIRM_SUCCESS:
			return {
				...state,
				status: 'waiting',
				items: action.payload.items,
				sortedItems: action.payload.items,
				step: action.payload.step,
			}

		case CalcActionTypes.SWITCH_TO_CONFIRM_ERROR:
			return {
				...state,
				status: 'error',
				error: action.payload,
			}

		case CalcActionTypes.ADD_INPUT_ITEM:
			return {
				...state,
				inputItems: [...state.inputItems, action.payload],
			}

		case CalcActionTypes.CHANGE_INPUT_ITEM:
			return {
				...state,
				inputItems: state.inputItems.map(item => {
					if (item.key === action.payload.key) {
						item.value = action.payload.value
					}
					return item
				}),
			}

		case CalcActionTypes.REMOVE_INPUT_ITEM:
			return {
				...state,
				inputItems: state.inputItems.filter(
					item => item.key !== action.payload
				),
			}

		case CalcActionTypes.SORT_ITEMS:
			return {
				...state,
				sortedItems: state.sortDirection
					? state.sortedItems.sort((a, b) => a - b)
					: state.sortedItems.sort((a, b) => b - a),
				sortDirection: !state.sortDirection,
			}

		case CalcActionTypes.SEARCH_ITEMS:
			return {
				...state,
				searchString: action.payload,
				sortedItems: state.items.filter(item =>
					matchCheck(item, action.payload)
				),
			}

		case CalcActionTypes.SWITCH_TO_DATAINPUT:
			return { ...state, status: 'loading', error: null }

		case CalcActionTypes.SWITCH_TO_DATAINPUT_SUCCESS:
			return {
				...state,
				status: 'waiting',
				items: action.payload.items,
				searchString: '',
				sortedItems: action.payload.items,
				step: action.payload.step,
				inputItems: [
					{ key: 'a13key1', value: '' },
					{ key: 'a13key2', value: '' },
				],
			}

		case CalcActionTypes.SWITCH_TO_DATAINPUT_ERROR:
			return {
				...state,
				status: 'error',
				error: action.payload,
			}

		case CalcActionTypes.SWITCH_TO_RESULT:
			return { ...state, status: 'loading', error: null }

		case CalcActionTypes.SWITCH_TO_RESULT_SUCCESS:
			return {
				...state,
				status: 'waiting',
				items: action.payload.items,
				searchString: '',
				sortedItems: action.payload.items,
				step: action.payload.step,
				summary: action.payload.summary,
			}

		case CalcActionTypes.SWITCH_TO_RESULT_ERROR:
			return {
				...state,
				status: 'error',
				error: action.payload,
			}

		case CalcActionTypes.RESET_TO_DATA_INPUT:
			return { ...state, status: 'loading', error: null }

		case CalcActionTypes.RESET_TO_DATA_INPUT_SUCCESS:
			return {
				...state,
				status: 'waiting',
				items: action.payload.items,
				sortedItems: action.payload.items,
				step: action.payload.step,
				summary: 0,
			}

		case CalcActionTypes.RESET_TO_DATA_INPUT_ERROR:
			return {
				...state,
				status: 'error',
				error: action.payload,
			}

		case CalcActionTypes.RESET_TO_NEW_CALCULATION:
			return { ...state, status: 'loading', error: null }

		case CalcActionTypes.RESET_TO_NEW_CALCULATION_SUCCESS:
			return {
				...state,
				status: 'waiting',
				items: action.payload.items,
				sortedItems: action.payload.items,
				step: action.payload.step,
				inputItems: [
					{ key: 'a13key1', value: '' },
					{ key: 'a13key2', value: '' },
				],
				summary: 0,
			}

		case CalcActionTypes.RESET_TO_NEW_CALCULATION_ERROR:
			return {
				...state,
				status: 'error',
				error: action.payload,
			}

		default:
			return state
	}
}
