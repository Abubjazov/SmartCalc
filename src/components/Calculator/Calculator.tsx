import { CalcDataConfirm } from '../CalcDataConfirm/CalcDataConfirm'
import { CalcDataInput } from '../CalcDataInput/CalcDataInput'
import './Calculator.scss'

export const Calculator = (): JSX.Element => {
	return (
		<div className='calc-widget'>
			<CalcDataInput />
			<CalcDataConfirm />
			{/*

			<div className='calc-result' style={{ display: 'none' }}>
				<h1>Результаты расчёта</h1>

				<ul>
					<li>12</li>
					<li>13</li>
					<li>14</li>
				</ul>

				<h2>39</h2>
				<div className='calc-result-footer'>
					<NavLink to='/smartcalc/calc'>Назад</NavLink>
					<NavLink to='/smartcalc/calc'>Новый расчёт</NavLink>
				</div>
			</div> */}
		</div>
	)
}
