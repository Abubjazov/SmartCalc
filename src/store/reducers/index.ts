import { combineReducers } from 'redux'
import { authReducer } from './authReducer'
import { calcReducer } from './calcReducer'

export const rootReducer = combineReducers({
	calc: calcReducer,
	auth: authReducer,
})

export type RootState = ReturnType<typeof rootReducer>
