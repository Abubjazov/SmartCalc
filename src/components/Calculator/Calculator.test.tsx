import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { BrowserRouter } from 'react-router-dom'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import { Calculator } from './Calculator'

import { inputItemsInitValue } from '../../store/reducers/calcReducer'

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

describe('Component: Calculator', () => {
	test('should render CalcDataInput with loading spinner', () => {
		const store = mockStore({
			auth: {
				token: 'testtoken',
				email: 'test@test.com',
			},
			calc: {
				step: 1,
				status: 'loading',
				error: null,
				inputItems: [...inputItemsInitValue],
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<Calculator />
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})

	test('should render CalcDataConfirm with loading spinner', () => {
		const store = mockStore({
			auth: {
				token: 'testtoken',
				email: 'test@test.com',
			},
			calc: {
				step: 2,
				status: 'loading',
				error: null,
				inputItems: [...inputItemsInitValue],
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<Calculator />
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})

	test('should render CalcDataResult with loading spinner', () => {
		const store = mockStore({
			auth: {
				token: 'testtoken',
				email: 'test@test.com',
			},
			calc: {
				step: 4,
				status: 'loading',
				error: null,
				inputItems: [...inputItemsInitValue],
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<Calculator />
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})

	test('should render CalcDataInput with error message', () => {
		const store = mockStore({
			auth: {
				token: 'testtoken',
				email: 'test@test.com',
			},
			calc: {
				step: 1,
				status: 'error',
				error: 'ERROR*Test Error String>>test',
				inputItems: [...inputItemsInitValue],
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<BrowserRouter>
					<Calculator />
				</BrowserRouter>
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})

	test('should render CalcDataConfirm with error message', () => {
		const store = mockStore({
			auth: {
				token: 'testtoken',
				email: 'test@test.com',
			},
			calc: {
				step: 2,
				status: 'error',
				error: 'ERROR*Test Error String>>test',
				inputItems: [...inputItemsInitValue],
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<BrowserRouter>
					<Calculator />
				</BrowserRouter>
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})

	test('should render CalcDataResult with error message', () => {
		const store = mockStore({
			auth: {
				token: 'testtoken',
				email: 'test@test.com',
			},
			calc: {
				step: 4,
				status: 'error',
				error: 'ERROR*Test Error String>>test',
				inputItems: [...inputItemsInitValue],
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<BrowserRouter>
					<Calculator />
				</BrowserRouter>
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})

	test('should render CalcDataInput', () => {
		const store = mockStore({
			auth: {
				token: 'testtoken',
				email: 'test@test.com',
			},
			calc: {
				step: 1,
				status: 'waiting',
				error: null,
				inputItems: [...inputItemsInitValue],
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<Calculator />
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})

	test('should render CalcDataConfirm', () => {
		const store = mockStore({
			auth: {
				token: 'testtoken',
				email: 'test@test.com',
			},
			calc: {
				step: 2,
				status: 'waiting',
				error: null,
				items: [1, 2, 3],
				sortedItems: [1, 2, 3],
				sortDirection: true,
				searchString: '',
				inputItems: [
					{ key: 'a13key1', value: '1' },
					{ key: 'a13key2', value: '2' },
					{ key: 'a13key3', value: '3' },
				],
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<Calculator />
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})

	test('should render CalcDataResult', () => {
		const store = mockStore({
			auth: {
				token: 'testtoken',
				email: 'test@test.com',
			},
			calc: {
				step: 4,
				status: 'waiting',
				error: null,
				items: [1, 2, 11],
				sortedItems: [1, 2, 11],
				sortDirection: true,
				searchString: '',
				inputItems: [
					{ key: 'a13key1', value: '1' },
					{ key: 'a13key2', value: '2' },
					{ key: 'a13key3', value: '11' },
				],
				summary: 14,
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<Calculator />
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})

	test('should render empty Calculator', () => {
		const store = mockStore({
			auth: {
				token: 'testtoken',
				email: 'test@test.com',
			},
			calc: {
				step: 3,
				status: 'waiting',
				error: null,
				inputItems: [...inputItemsInitValue],
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<Calculator />
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})
})
