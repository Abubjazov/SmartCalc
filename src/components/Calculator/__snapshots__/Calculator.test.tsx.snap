// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Calculator should render CalcDataConfirm 1`] = `
<DocumentFragment>
  <div
    class="calc-widget"
  >
    <p
      class="calc-widget-p"
    >
      Привет, Test@t***.com
    </p>
    <div
      class="calc-data-confirm"
    >
      <h1>
        Подтвердите данные
      </h1>
      <div
        class="calc-data-confirm-search"
      >
        <input
          placeholder="Поиск"
          type="text"
          value=""
        />
        <div
          class="sort null"
          tabindex="0"
        >
          ^
        </div>
      </div>
      <ul>
        <li>
          1
        </li>
        <li>
          2
        </li>
        <li>
          3
        </li>
      </ul>
      <div
        class="calc-data-confirm-footer"
      >
        <button>
          Назад
        </button>
        <button>
          Подтверждаю
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Component: Calculator should render CalcDataInput 1`] = `
<DocumentFragment>
  <div
    class="calc-widget"
  >
    <p
      class="calc-widget-p"
    >
      Привет, Test@t***.com
    </p>
    <div
      class="calc-data-input"
    >
      <h1>
        Введите данные
      </h1>
      <div
        class="calc-input"
      >
        <input
          class="null"
          placeholder="Введите число"
          type="text"
          value=""
        />
      </div>
      <div
        class="calc-input"
      >
        <input
          class="null"
          placeholder="Введите число"
          type="text"
          value=""
        />
      </div>
      <div
        class="add-item"
        tabindex="0"
      >
        +
      </div>
      <button>
        Далее
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Component: Calculator should render CalcDataResult 1`] = `
<DocumentFragment>
  <div
    class="calc-widget"
  >
    <p
      class="calc-widget-p"
    >
      Привет, Test@t***.com
    </p>
    <div
      class="calc-data-result"
    >
      <h1>
        Результаты расчёта
      </h1>
      <ul>
        <li
          class=""
        >
          1
        </li>
        <li
          class=""
        >
          2
        </li>
        <li
          class="green"
        >
          11
        </li>
      </ul>
      <h2>
        Сумма: 14
      </h2>
      <div
        class="calc-data-result-footer"
      >
        <button>
          Ввод данных
        </button>
        <button>
          Новый расчёт
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Component: Calculator should render Calculator with error message 1`] = `
<DocumentFragment>
  <div
    class="calc-widget"
  >
    <div
      class="error-message"
      role="alert"
    >
      <p
        class="error-message-p"
        tabindex="0"
      >
        ERROR
      </p>
      <p
        class="error-message-p"
        tabindex="0"
      >
        Test Error String
      </p>
      <a
        class="error-message-a"
        href="/smartcalc/"
      >
        Вернуться на главную
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Component: Calculator should render Calculator with loading spinner 1`] = `
<DocumentFragment>
  <div
    class="calc-widget"
  >
    <svg
      class="spinner"
      height="200"
      preserveAspectRatio="xMidYMid"
      viewBox="0 0 100 100"
      width="200"
    >
      <path
        d="M24.3 30C11.4 30 5 43.3 5 50s6.4 20 19.3 20c19.3 0 32.1-40 51.4-40 C88.6 30 95 43.3 95 50s-6.4 20-19.3 20C56.4 70 43.6 30 24.3 30z"
        fill="none"
        stroke="#cccccc"
        stroke-dasharray="42.76482137044271 42.76482137044271"
        stroke-linecap="round"
        stroke-width="8"
      >
        <animate
          attributeName="stroke-dashoffset"
          dur="1s"
          keyTimes="0;1"
          repeatCount="indefinite"
          values="0;256.58892822265625"
        />
      </path>
    </svg>
  </div>
</DocumentFragment>
`;
